ghost/core/core/server/services/link-tracking/PostLinkRepository.js

```diff
@@ -48,6 +62,116 @@ module.exports = class PostLinkRepository {
         return result;
     }
 
+    /**
+     * validate and manage the new redirect url in filter
+     * `to` url needs decoding and transformation to relative url for comparision
+     * @param {string} filter
+     * @returns {Object} parsed filter
+     * @throws {errors.BadRequestError}
+     */
+    #parseLinkFilter(filter) {
+        // decode filter to manage any encoded uri components
+        filter = decodeURIComponent(filter);
+
+        try {
+            const filterJson = nql(filter).parse();
+            const postId = filterJson?.$and?.[0]?.post_id;
+            const redirectUrl = new URL(filterJson?.$and?.[1]?.to);
+            if (!postId || !redirectUrl) {
+                throw new errors.BadRequestError({
+                    message: tpl(messages.invalidFilter)
+                });
+            }
+            return {
+                postId,
+                redirectUrl
+            };
+        } catch (e) {
+            throw new errors.BadRequestError({
+                message: tpl(messages.invalidFilter),
+                context: e.message
+            });
+        }
+    }
+
+    #getRedirectLinkWithAttribution({newLink, oldLink, postId}) {
```

Comment:
I don't think this is the correct place to put this logic, this class should only be concerned with the persistence and retrieval of data from the DB - but the params here are very much application/feature code - and should probably live in the `@tryghost/link-tracking` package

Reply:
@allouis Updated to move logic out to `LinkTrackingService`, and the `PostLinkRepository` now gets the link ids to update along with the new data so it doesn't has any logic.